### DISCUSSION TITLE
Generating NPM package files

### DISCUSSION THREAD

**Discussion by gunnartorfis** [orval-labs/orval#240] [Ideas]
We would love to be able to automatically generate NPM package files, package.json and .npmrc. 

In our CircleCI pipeline, we'd be releasing a couple of versions of our client (@semver, @beta, @next), etc.
Is this something that would make sense for Orval to do our should we just create a script ourselves that does this?

**Comment by anymaniax** [comment]
Hello, like that I would say it's not in the scope of Orval but can you provide more detail about what you need or an example of the generation that you would need?

**Reply by gunnartorfis** [reply]
Thanks for the reply. 
I implemented it in our project already but I'd be happy to make a PR myself to Orval. 

Our use case is as follows. We have our API and then a couple of frontend projects. 
Instead of them applying Orval in every single frontend, we wanted to move our Orval generation to the API itself and rather publish it to NPM. So in our CI/CD pipeline, the last step is to publish the Orval generated client. We have tags for some of our development environment branches, etc. 

High level description of the script: 
- Copy orval.config.js and .npmrc to a new directory that is to be published
- Generate the Orval (react-query in our case) client 
- Initialize a new NPM project 
- Modify generated package.json with semver versioning and main entry 
- Publish the package

**Reply by anymaniax** [reply]
I think it's out of the scope but you can maybe create an example with your script that we can add with the others?

**Comment by Talor-A** [comment]
@gunnartorfis I'm trying to accomplish this right now, it would be great if you could share some snippets here or post an example repo. the solution I went with was creating a "generator" package, with a nested `generated-client` folder inside it (with its own package.json, etc). the generated files go into `generated-client/src`, and then I run `npm version <myversion>` inside the client and publish it. I'm still working on this and running into some small snags, so another example would be great.

**Reply by gunnartorfis** [reply]
My solution is unfortunately a bit specific to our project. We use CircleCI to generate the client, Semver versioning for pushes on main and then a tag for each branch with the name of the branch. 

In our .circleci/scripts/client_generations/typescript folder, I have the following files:
- .npmrc for publishing our private NPM package
- orval.config.js
- README.md (for npmjs.org, including installation, usage, etc)
- typescript.sh - here's most of the logic

**typescript.sh**
```bash
#!/bin/bash

if [[ $# -eq 0 ]] ; then
    echo 'Missing argument: version'
    exit 0
fi

VERSION=$1
NPM_VERSION=$VERSION
NPM_TAG="latest"

if [[ "$CIRCLE_BRANCH" != "main" ]]
then
    NPM_TAG=$CIRCLE_BRANCH
fi

# Install Yarn
if ! command -v yarn &> /dev/null
then
    echo "yarn could not be found, installing"
    npm install --global yarn
else
    echo "yarn installed, skipping"
fi

# Install Orval CLI
if ! command -v orval &> /dev/null
then
    echo "orval could not be found, installing"
    yarn global add orval
else
    echo "orval installed, skipping"
fi

cd .circleci/scripts/client_generations/typescript

# This is the place where the client gets generated, remove it if there's one there already.
# More work is done later in the script in this folder.
rm -rf ../../../../clients/ts/noona-api

# Generate client based on orval.config.js
orval

# Move .npmrc to client directory to be able to publish
cp .npmrc ../../../../clients/ts/noona-api
cp README.md ../../../../clients/ts/noona-api

# CD into client directory and init a NPM package
cd ../../../../clients/ts/noona-api
yarn config set init-author-name XXXX
yarn config set init-version ${NPM_VERSION}
yarn init --yes

# Install JQ CLI
# JQ is used for patching the package.json later on
if ! command -v jq &> /dev/null
then
    echo "jq could not be found, installing"
    if [ "$(uname)" == "Darwin" ]; then
        # Do something under Mac OS X platform       
        brew install jq
    elif [ "$(expr substr $(uname -s) 1 5)" == "Linux" ]; then
        # Do something under GNU/Linux platform
        apt-get -y install jq
    fi
else
    echo "jq installed, skipping"
fi

yarn add -D typescript
yarn add -D axios
yarn add -D react-query

yarn global add npx
npx tsc --init

# The part from here is a bit specific to how you want your NPM package to be structured. This is my first published package with Typescript support so I'm sure there are better ways to structure it. 

# Rename package.json from noona-api to @noonahq/noona-api
jq -M '. + {name:"@noonahq/noona-api", main:"dist/index.js", types:"types/index.d.ts"}' package.json > new.package.json
#jq '.name="@noonahq/noona-api" | .main="noona-api.ts"' package.json > new.package.json
mv new.package.json package.json

npx tsc --declaration --noImplicitAny false --outDir dist

mkdir types
mv dist/index.d.ts types/index.d.ts
echo "export * as Models from './models'" >> types/index.d.ts

mv dist/models types
rm types/models/*.js

rm -rf models
rm index.ts

echo "node_modules" > .npmignore

yarn remove axios
yarn remove react-query
yarn add -P axios
yarn add -P react-query
rm -rf node_modules/react-query

npm publish --tag $NPM_TAG
```

Then in our CircleCI's .circleci/config.yml, we add a step that runs the typescript.sh file to generate & publish the client.
